rbd.c

```diff
@@ -76,6 +77,89 @@ struct rbd_aio_cb {
 	char *bounce_buffer;
 };
 
+#ifdef LIBRADOS_SUPPORTS_SERVICES
+
+static void tcmu_rbd_service_status_update(struct tcmu_device *dev,
+					   bool has_lock)
+{
+	int ret;
+	char *status_buf = NULL;
+	ret = asprintf(&status_buf, "%s%c%s%c", "lock_owner", '\0',
+		       has_lock ? "true" : "false", '\0');
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not allocate status buf. Service will not be updated.\n");
+		return;
+	}
+
+	ret = rados_service_update_status(state->cluster, status_buf);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not update service status. (Err %d\n",
+			     ret);
+	}
+
+	free(status_buf);
+}
+
+static int tcmu_rbd_service_register(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	struct utsname u;
+	char *daemon_buf = NULL;
+	char *metadata_buf = NULL;
+	int ret;
+
+	ret = uname(&u);
+	if (ret < 0) {
+		ret = -errno;
+		tcmu_dev_err(dev, "Could not query uname. (Err %d)\n", ret);
+		return ret;
+	}
+
+	ret = asprintf(&daemon_buf, "%s:%s/%s",
+		       u.nodename, state->pool_name, state->image_name);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not allocate daemon buf.\n");
+		return -ENOMEM;
+	}
+
+	ret = asprintf(&metadata_buf, "pool_name%c%s%cimage_name%c%s%c",
+		       '\0', state->pool_name, '\0',
+		       '\0', state->image_name, '\0');
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not allocate metadata buf.\n");
+		ret = ENOMEM;
+		goto free_daemon_buf;
+	}
+
+	ret = rados_service_register(state->cluster, "tcmu-runner",
+				     daemon_buf, metadata_buf);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not register service to cluster. (Err %d)\n",
+			     ret);
+	}
+
+	free(metadata_buf);
+free_daemon_buf:
+	free(daemon_buf);
+	return ret;
+}
+
+#else
+
+static int tcmu_rbd_service_register(struct tcmu_device *dev)
+{
+	/* Ignorable. Just log in dbg mode just in case. */
+	tcmu_dev_dbg(dev, "Could not register service to ceph cluster.\n");
```

Comment:
Nit: should this message be something like "service registration not supported"?